### PR TITLE
Transition to valtrees pt1

### PR DESCRIPTION
**Pull request by b-naber**
Compartmentalising https://github.com/rust-lang/rust/pull/96591 as much as possible.

r? @oli-obk

### CODE REVIEW

File: compiler/rustc_const_eval/src/const_eval/valtrees.rs [hunk 1]
```diff
@@ -223,8 +241,11 @@ fn create_pointee_place<'tcx>(
             .unwrap();
         debug!(?ptr);
 
-        let mut place = MPlaceTy::from_aligned_ptr(ptr.into(), layout);
-        place.meta = MemPlaceMeta::Meta(Scalar::from_u64(num_elems as u64));
```
Comment (RalfJung) [inline compiler/rustc_const_eval/src/const_eval/valtrees.rs:227]: As mentioned before, please also remove the `impl DerefMut for MPlaceTy`

**Comment by b-naber** [comment]
I'm not sure why I get the changes in the submodules. Something must have gone wrong during a rebase, I had some conflicts in submodules and used `get reset master submodule_path` to resolve those, which might have been the wrong way to fix those. Does anybody know how to fix this?

**Comment by RalfJung** [comment]
Figure out in which commit the submodule change is, and then use `rebase --interactive` to `edit` that commit and get rid of the submodule change.

**Comment by b-naber** [comment]
Updated the submodule and squashed with the commit that previously had the submodule change. Still have a change in the submodule's `Cargo.lock` file according to the diff though.

**Review by oli-obk — COMMENTED** [review]
I went afk mid review and forgot to send it. Will continue on Monday.

**Comment by oli-obk** [comment]
needs a rebase.

Also you got a submodule update somewhere in there, please make sure to remove it during the rebase

**Comment by oli-obk** [comment]
r=me after that

File: compiler/rustc_const_eval/src/const_eval/valtrees.rs [hunk 2]
```diff
@@ -1,124 +1,140 @@
...
     ecx: &CompileTimeEvalContext<'tcx, 'tcx>,
     place: &MPlaceTy<'tcx>,
-) -> Option<ty::ValTree<'tcx>> {
-    let n = place
-        .len(&ecx.tcx.tcx)
-        .unwrap_or_else(|_| panic!("expected to use len of place {:?}", place));
-    let branches = (0..n).map(|i| {
+    num_nodes: &mut usize,
+) -> ValTreeCreationResult<'tcx> {
+    let n = place.len(&ecx.tcx.tcx).expect(&format!("expected to use len of place {:?}", place));
```
Comment (klensy) [inline compiler/rustc_const_eval/src/const_eval/valtrees.rs:59]: Uhh, you're basically reverted #96982.

**Comment by b-naber** [comment]
@oli-obk Can you take another look, please?

**Comment by oli-obk** [comment]
cachegrind on the `unicode-normalization` regression:

```
-97,176,447  ???:rustc_mir_build::thir::pattern::compare_const_vals
 72,480,660  ???:<rustc_middle::mir::interpret::value::ConstValue>::try_to_bits
 54,411,647  ???:<rustc_mir_build::build::Builder>::match_simplified_candidates
-37,703,135  ???:<rustc_mir_build::build::Builder>::test_candidates_with_or
 20,818,165  ???:<rustc_middle::mir::ConstantKind>::ty
 18,643,459  ???:<rustc_mir_build::build::Builder>::values_not_contained_in_range
  9,696,591  ???:<rustc_mir_build::build::Builder>::simplify_candidate
 -4,265,882  ???:rustc_mir_build::build::expr::as_constant::lit_to_mir_constant
  2,721,211  ???:<rustc_mir_build::build::Builder>::as_constant
```

could probably benefit from some targeted `#[inline(always)]` instructions (`try_to_bits` and `ConstantKind::ty` maybe?)

**Pull request merged**